addons/mail/controllers/discuss_controller.py

```diff
@@ -0,0 +1,284 @@
+            ('message_type', '!=', 'user_notification'),
+        ]
+
+        return self._fetch_messages(domain, min_id, max_id, limit)
+    
+    @http.route('/mail/channel/<int:channel_id>/mark_as_seen', type='json', auth='public')
+    def mark_channel_as_seen(self, channel_id, last_message_id, **kwargs):
+        channel_partner = self._get_channel_partner(channel_id)
+        if not channel_partner:
+            raise NotFound()
+
+        return channel_partner.channel_id.channel_seen(last_message_id)
+
+    @http.route('/mail/init_messaging', type='json', auth='user')
+    def mail_init_messaging(self, **kwargs):
+        return request.env.user._init_messaging()
+
+    @http.route('/mail/init_messaging_public/<int:channel_id>', type='json', auth='public')
+    def mail_init_messaging_public(self, channel_id, **kwargs):
+        channel_partner = self._get_channel_partner(channel_id)
+        if not channel_partner:
+            raise NotFound()
+
+        return channel_partner.env.user._init_messaging()
+
+    @http.route('/mail/message/history', type='json', auth='user')
+    def fetch_messages_from_history(self, min_id=False, max_id=False, limit=30, **kwargs):
+        domain = [('needaction', '=', False)]
+        return self._fetch_messages(domain, min_id, max_id, limit)
+
+    @http.route('/mail/message/inbox', type='json', auth='user')
+    def fetch_messages_from_inbox(self, min_id=False, max_id=False, limit=30, **kwargs):
+        domain = [('needaction', '=', True)]
+        return self._fetch_messages(domain, min_id, max_id, limit)
+
+    @http.route('/mail/message/starred_messages', type='json', auth='user')
+    def fetch_messages_from_starred(self, min_id=False, max_id=False, limit=30, **kwargs):
+        domain = [('starred_partner_ids', 'in', [request.env.user.partner_id.id])]
+        return self._fetch_messages(domain, min_id, max_id, limit)
+
+    @http.route('/mail/thread/<string:thread_model>/<int:thread_id>/fetch_messages', type='json', auth='user')
+    def fetch_messages_from_thread(self, thread_model, thread_id, max_id=False, min_id=False, limit=30, **kwargs):
+        domain = [
+            ('res_id', '=', thread_id),
+            ('model', '=', thread_model),
+            ('message_type', '!=', 'user_notification'),
+        ]
+
+        return self._fetch_messages(domain, min_id, max_id, limit)
+    
+    @http.route('/mail/<string:thread_model>/<int:thread_id>/post_message', type='json', auth='public')
+    def post_message(self, thread_model, thread_id, postData, **kwargs):
+        if thread_model == 'mail.channel':
+            channel_partner = self._get_channel_partner(thread_id)
+            if not channel_partner:
+                raise NotFound()
+            return channel_partner.channel_id.message_post(**postData)
+
+        if not request.session.uid:
+            raise NotFound()
+
+        return request.env[thread_model].search([('id', '=', thread_id)], limit=1).message_post(**postData)
```

Comment:
.search([('id', '=', thread_id)], limit=1) => .browse(thread_id)

Reply:
There is a small possible difference in semantics: `search` will return an empty recordset if the record is invalid. With `browse` you have to follow with `exists`.